### PR TITLE
[router] Prevent base url from being appended to external links

### LINKED ISSUE #26566: ExternalLink adds client route
### Minimal reproducible example

https://github.com/rocket-meals/expo-baseUrl-reload-error

### Which package manager are you using? (Yarn is recommended)

npm

### If the issue is web-related, please select the bundler (`web.bundler` in the `app.json`)

metro

### Summary

```
let url = "https://rocket-meals.de/demo/api/auth/..."
```

```
<ExternalLink target={"_self"} href={url} style={styles.link}>
            <Text style={styles.linkText}>{"Login with: "+provider+"\n"}</Text>
            <Text style={styles.linkText}>{"URL: "+url}</Text>
        </ExternalLink>
```

So what i get displayed is:

```
Login with: google
URL: https://rocket-meals.de/demo/api/auth/login/google?...
```

But when hovering with mouse over the link:

```
http://localhost:3000/rocket-mealshttps://rocket-meals.de/demo/api/auth/login/google...
```

It seems to add the current url of the frontend just before. This only happens in production. In dev mode it works as expected:

In production mode:
Bildschirmfoto 2024-01-21 um 20 00 09

Seems to be related to: https://github.com/expo/expo/issues/26565

### Environment

System:
  OS: macOS 13.4.1
  Shell: 5.9 - /bin/zsh
Binaries:
  Node: 16.20.2 - ~/.nvm/versions/node/v16.20.2/bin/node
  Yarn: 1.22.21 - /usr/local/bin/yarn
  npm: 8.19.4 - ~/.nvm/versions/node/v16.20.2/bin/npm
  Watchman: 2023.11.27.00 - /usr/local/bin/watchman
Managers:
  CocoaPods: 1.14.3 - /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
IDEs:
  Android Studio: 2023.1 AI-231.9392.1.2311.11076708
  Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
npmGlobalPackages:
  eas-cli: 6.0.0
Expo Workflow: managed

### PR DESCRIPTION
**Pull request by 6TELOIV**
# Why

Fixes #26566

The bug is that `EXPO_BASE_URL` is added to hrefs that are not part of the app's navigation structure. For example, `https://www.example.com/` becomes `/base_urlhttps://www.example.com/`)

# How

Checking that the `href` doesn't include a schema nor authority ensures that we don't add base url to links which should not have it (links like `https://www.example.com/` or `//www.example.com/`).

# Test Plan

Added 3 tests checking base url appending behavior in network relative uris, path uris, and absolute uris. This should cover the URI types that we don't want to append base url to.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/link/useLinkToPathProps.tsx [hunk 1]
```diff
@@ -52,9 +73,23 @@ export default function useLinkToPathProps({ href, ...options }: UseLinkToPathPr
...
 
+  const baseAppendedStrippedHref = React.useMemo(() => {
+    const strippedHref = stripGroupSegmentsFromPath(href);
+    // Ensure there's always a value for href.
+    if (!strippedHref) {
+      return appendBaseUrl('/');
+    }
+
+    // Append base url only if needed.
+    if (shouldAppendBaseUrl(strippedHref)) {
```
Comment (EvanBacon) [inline packages/expo-router/src/link/useLinkToPathProps.tsx:84]: Use `shouldLinkExternally` https://github.com/expo/expo/blob/df05000d67f74aae9c5e03a5d9c5b95807220025/packages/expo-router/src/utils/url.ts#L16

Comment (6TELOIV) [reply packages/expo-router/src/link/useLinkToPathProps.tsx]: Changed! Didn't spot that one existing already.

**Comment by 6TELOIV** [comment]
I think I did not run `yarn build`, causing the SDK check to fail. I'm updating the pull request now, my appoligies! This is my first contribution 😛

**Comment by marklawlor** [comment]
Same protocol hrefs don't make sense on a native device, so being explicit is better.

But I'm sure there is some edge case where you might need it, so let me discuss this PR https://github.com/expo/expo/pull/31646 with the team first.

**Comment by 6TELOIV** [comment]
Absolutely; makes sense.

**Comment by 6TELOIV** [comment]
I realized now that when changing to `shouldLinkExternally` I was appending baseUrl for external links only, when baseUrl should only be appended to non-external links. I forgot to build the package before committing, so I'm doing so now and hopefully will see green lights across the board.

**Comment by 6TELOIV** [comment]
I don't have access to a linux env where I can rebuild the package. I will try again later when I'm on my personal device.

**Comment by 6TELOIV** [comment]
Whew, okay. I think I've got it all working now. Tests were passing on my local, and I properly built the packages. 🤞🤞🤞

**Comment by 6TELOIV** [comment]
Yay! Green lights 😄

**Review by marklawlor — APPROVED** [review]

**Pull request merged**